src/main/java/io/lettuce/authx/TokenBasedRedisCredentialsProvider.java

```diff
@@ -0,0 +1,101 @@
+import io.lettuce.core.StreamingCredentialsProvider;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.core.publisher.Sinks;
+import redis.clients.authentication.core.Token;
+import redis.clients.authentication.core.TokenAuthConfig;
+import redis.clients.authentication.core.TokenListener;
+import redis.clients.authentication.core.TokenManager;
+
+public class TokenBasedRedisCredentialsProvider implements StreamingCredentialsProvider {
+
+    private final TokenManager tokenManager;
+
+    private final Sinks.Many<RedisCredentials> credentialsSink = Sinks.many().replay().latest();
+
+    public TokenBasedRedisCredentialsProvider(TokenAuthConfig tokenAuthConfig) {
+        this(new TokenManager(tokenAuthConfig.getIdentityProviderConfig().getProvider(),
+                tokenAuthConfig.getTokenManagerConfig()));
+
+    }
+
+    public TokenBasedRedisCredentialsProvider(TokenManager tokenManager) {
+        this.tokenManager = tokenManager;
+        initializeTokenManager();
+    }
+
+    /**
+     * Initialize the TokenManager and subscribe to token renewal events.
+     */
+    private void initializeTokenManager() {
+        TokenListener listener = new TokenListener() {
+
+            @Override
+            public void onTokenRenewed(Token token) {
+                String username = token.tryGet("oid");
+                char[] pass = token.getValue().toCharArray();
+                RedisCredentials credentials = RedisCredentials.just(username, pass);
+                credentialsSink.tryEmitNext(credentials);
+            }
+
+            @Override
+            public void onError(Exception exception) {
+                credentialsSink.tryEmitError(exception);
+            }
+
+        };
+
+        try {
+            tokenManager.start(listener, false);
+        } catch (Exception e) {
+            credentialsSink.tryEmitError(e);
+        }
+    }
+
+    /**
+     * Resolve the latest available credentials as a Mono.
+     * <p>
+     * This method returns a Mono that emits the most recent set of Redis credentials.
+     * The Mono will complete once the credentials are emitted.
+     * If no credentials are available at the time of subscription, the Mono will wait until
+     * credentials are available.
+     *
+     * @return a Mono that emits the latest Redis credentials
+     */
+    @Override
+    public Mono<RedisCredentials> resolveCredentials() {
+
+        return credentialsSink.asFlux().next();
+    }
+
+    /**
+     * Expose the Flux for all credential updates.
+     * <p>
+     * This method returns a Flux that emits all updates to the Redis credentials.
+     * Subscribers will receive the latest credentials whenever they are updated.
+     * The Flux will continue to emit updates until the provider is shut down.
+     *
+     * @return a Flux that emits all updates to the Redis credentials
+     */
+    @Override
+    public Flux<RedisCredentials> credentials() {
```

Comment:
`TokenBasedRedisCredentialsProvider` looks nice and clean overall, and does most of the job as i understand. Only thing i got around is that i believe it would do auth twice with the same initial token;
 one with a call to

```java
 public Mono<RedisCredentials> resolveCredentials() {
 ... 
```

and the other with a call to
```java
public Flux<RedisCredentials> credentials() {
...
```

Reply:
Yes, It will do it twice.
I was thinking to skip first token when subscribing since it was used during the handshake and this way avoid the second auth, but my thinking is that between the handshake and connection activation, there might be token update that we miss (it should be very rare but still possible) and end up with expired token. Also since we are targeting long running connections where we have number of reconnects, I think it is safer to leave the initial duplication of auth to avoid complex sync.

Reply:
this makes me question the way handshake and authentication works and how the entities passed over between actors. i have a gut feeling that authentication handler should be handling authentications;
may be hand shake sequence(`RedisHandshake`) should initiate and utilize the `AuthenticationHandler` , how that sounds ?

Reply:
We might consider also that approach but  handshake performs also other tasks (sets protocol, client name and auth) in single HELLO command.
So to me it make sense to keep them separated, 
handshake    -> initial connection setup, 
auth handler -> if enabled is responsible only for keeping the channel authenticated with up-to date token.

Reply:
TBH i dont know; this might also mean a reasonable amount of change on the PR and i am not very sure of it s the right thing to do. 
So let's just do nothing but think about it for now and may be we will reconsider changing the approach in another iteration if there is one.